### PR TITLE
[FIX] mail

### PR DESCRIPTION
**Pull request by tde-banana-odoo**
Task: https://www.odoo.com/web#id=28652&view_type=form&model=project.task&action=333&active_id=248&menu_id=4720

### CODE REVIEW

File: addons/crm/controllers/main.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+import werkzeug
+from werkzeug import url_encode
+
+from odoo.addons.mail.controllers.main import MailController
+from odoo import http
+from odoo.http import request
+
+
+class HrHolidaysController(http.Controller):
```
Comment (dreispt) [inline addons/crm/controllers/main.py]: HrHolidaysController in CRM?

Comment (tde-banana-odoo) [inline addons/crm/controllers/main.py]: Sometimes I have a feeling that someone spies me :) . Anyway this is just a WIP. But you are right and I modified it. Copy and paste magic.

File: addons/crm/controllers/main.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+from openerp.http import request
+
+
+class CrmController(http.Controller):
+
+    @http.route('/lead/case_mark_won', type='http', auth='user', methods=['POST'], csrf=False)
+    def crm_lead_case_mark_won(self, model, res_id):
+        if model not in request.env:
+            return MailController._redirect_to_messaging()
+        Model = request.env[model]
```
Comment (KangOl) [inline addons/crm/controllers/main.py]: model should be hard-coded to `crm.lead`

Method should only take lead id as param.

Comment (tde-banana-odoo) [inline addons/crm/controllers/main.py]: Stop trying the new review mechanism on my PRs ! Goujat !

File: addons/mail/controllers/main.py [hunk 3]
```diff
@@ -84,8 +97,8 @@ def read_subscription_data(self, res_model, res_id, follower_id=None):
         subtypes_list = sorted(subtypes_list, key=itemgetter('parent_model', 'res_model', 'internal', 'sequence'))
         return subtypes_list
 
-    @http.route('/mail/view', type='http', auth='none')
-    def mail_action_view(self, model=None, res_id=None, message_id=None):
+    @http.route('/mail/view', type='http', auth='none', methods=['GET'])
+    def mail_action_view(self, token, model=None, res_id=None, message_id=None):
```
Comment (KangOl) [inline addons/mail/controllers/main.py]: Existing route should not change.
Already sent links must continue to works

Comment (tde-banana-odoo) [inline addons/mail/controllers/main.py]: You think keeping security holes is necessary just for some email links ?

Comment (KangOl) [inline addons/mail/controllers/main.py]: The route `/mail/view` won't let you view the record if you don't have access.

Comment (tde-banana-odoo) [inline addons/mail/controllers/main.py]: Oh that one. Thought you were talking about all modified routes. Indeed this one can probably be kept.

Comment (KangOl) [inline addons/mail/controllers/main.py]: Actually, other remaining route are safe.
Only `/mail/assign` may be problematic.

**But** this route wasn't used due to bug in `_notification_link_helper` which generated a `/mail/view` route.
I propose to use a specific controller for the only places that was using it (project.task/project.issue)

File: addons/mail/controllers/main.py [hunk 4]
```diff
@@ -24,6 +29,15 @@ def _redirect_to_messaging(cls):
         return werkzeug.utils.redirect(url)
 
     @classmethod
+    def _check_token(cls, token):
+        base_link = request.httprequest.environ['PATH_INFO']
+        params = request.params
+        params.pop('token', '')
```
Comment (KangOl) [inline addons/mail/controllers/main.py]: You should make a copy of params before altering it.

File: addons/mail/controllers/main.py [hunk 5]
```diff
@@ -24,6 +29,15 @@ def _redirect_to_messaging(cls):
         return werkzeug.utils.redirect(url)
 
     @classmethod
+    def _check_token(cls, token):
+        base_link = request.httprequest.environ['PATH_INFO']
+        params = request.params
```
Comment (KangOl) [inline addons/mail/controllers/main.py]: Better make a copy of params before altering it (next line)

**Pull request closed without merging**